src/compiler/performanceCore.ts

```diff
@@ -0,0 +1,79 @@
+/*@internal*/
+namespace ts {
+    export interface PerformanceHooks {
+        performance: Performance;
+        PerformanceObserver: PerformanceObserverConstructor;
+    }
+
+    export interface Performance {
+        clearMarks(name?: string): void;
+        mark(name: string): void;
+        measure(name: string, startMark?: string, endMark?: string): void;
+        now(): number;
+        timeOrigin: number;
+    }
+
+    export interface PerformanceEntry {
+        name: string;
+        entryType: string;
+        startTime: number;
+        duration: number;
+    }
+
+    export interface PerformanceObserverEntryList {
+        getEntries(): PerformanceEntryList;
+        getEntriesByName(name: string, type?: string): PerformanceEntryList;
+        getEntriesByType(type: string): PerformanceEntryList;
+    }
+
+    export interface PerformanceObserver {
+        disconnect(): void;
+        observe(options: { entryTypes: readonly string[] }): void;
+    }
+
+    export type PerformanceObserverConstructor = new (callback: (list: PerformanceObserverEntryList, observer: PerformanceObserver) => void) => PerformanceObserver;
+    export type PerformanceEntryList = PerformanceEntry[];
+
+    declare const performance: Performance | undefined;
+    declare const PerformanceObserver: PerformanceObserverConstructor | undefined;
+
+    function tryGetWebPerformanceHooks(): PerformanceHooks | undefined {
+        if (typeof performance === "object" &&
+            typeof performance.timeOrigin === "number" &&
+            typeof performance.clearMarks === "function" &&
+            typeof performance.mark === "function" &&
+            typeof performance.measure === "function" &&
+            typeof performance.now === "function" &&
+            typeof PerformanceObserver === "function") {
+            return {
+                performance,
+                PerformanceObserver
+            };
+        }
+    }
+
+    function tryGetNodePerformanceHooks(): PerformanceHooks | undefined {
+        if (typeof module === "object" && typeof require === "function") {
+            try {
+                return require("perf_hooks") as typeof import("perf_hooks");
+            }
+            catch {
+            }
+        }
+    }
+
+    const nativePerformanceHooks = tryGetWebPerformanceHooks() || tryGetNodePerformanceHooks();
+
+    export function tryGetNativePerformanceHooks() {
+        return nativePerformanceHooks;
+    }
+
+    /** Gets a timestamp with (at least) ms resolution */
+    export const timestamp = (() => {
+        if (nativePerformanceHooks) {
+            const performance = nativePerformanceHooks.performance;
+            return () => performance.now();
+        }
+        return Date.now ? Date.now : () => +(new Date());
+    })();

```

Comment:
It's not a test on every call. `timestamp` is assigned the result of an IIFE (Immediately Invoked Function Expression) so as not to leak the `performance` variable to the outer scope (and to reduce the number of property access steps for the actual callback.  Without the IIFE, we'd be writing this instead:

```ts
export const timestamp = 
    nativePerformanceHooks ? () => nativePerformanceHooks!.performance.now() :
    Date.now ? Date.now :
    () => +(new Date());
```
Which means two property lookups for each call to `timestamp` (first for "performance", then for "now").

Unfortunately, this won't work:

```ts
export const timestamp = 
    nativePerformanceHooks ? nativePerformanceHooks.performance.now :
    Date.now ? Date.now :
    () => +(new Date());
```

Because the Web Performance API requires the receiver for `now` be correct:

![image](https://user-images.githubusercontent.com/3902892/93401395-f163e680-f836-11ea-9d7d-e44a794709eb.png)

Reply:
I simplified this a bit to store a `nativePerformance` so as not to clobber the global `performance`, and updated the `timestamp` definition to no longer use an IIFE.

Reply:
(Bah, I think that I got confused with the IIFE...  Sorry.)